docs/core/encoding.md

```diff
@@ -1,44 +1,79 @@
+Note, the Amino-based types may slowly be phased-out in the future so developers
+should take note to use the protobuf message definitions where possible.
+
+### Amino
+
+Every module uses an Amino codec to serialize types and interfaces. This codec typically
+has types and interfaces registered in that module's domain only (e.g. messages),
+but there are exceptions like `x/gov`. Each module exposes a `RegisterCodec` function
+that allows a user to provide a codec and have all the types registered. An application
+will call this method for each necessary module.
+
+Where there is no protobuf-based type definition for a module, Amino is used to encode
+and decode raw wire bytes to the concrete type or interface:
 
 ```go
-// typical pattern to marshal an object to []byte before storing it
-bz := keeper.cdc.MustMarshalBinaryBare(object)
-
-//typical pattern to unmarshal an object from []byte when retrieving it
-keeper.cdc.MustUnmarshalBinaryBare(bz, &object)
+bz := keeper.cdc.MustMarshalBinaryBare(typeOrInterface)
+keeper.cdc.MustUnmarshalBinaryBare(bz, &typeOrInterface)
 ```
 
-Alternatively, it is possible to use `MustMarshalBinaryLengthPrefixed`/`MustUnmarshalBinaryLengthPrefixed` instead of `MustMarshalBinaryBare`/`MustUnmarshalBinaryBare` for the same encoding prefixed by a `uvarint` encoding of the object to encode. 
+Note, there are length-prefixed variants of the above functionality and this is
+typically used for when the data needs to be streamed or grouped together
+(e.g. `ResponseDeliverTx.Data`)
 
-Another important use of the `codec` is the encoding and decoding of [transactions](./transactions.md). Transactions are defined at the Cosmos SDK level, but passed to the underlying consensus engine in order to be relayed to other peers. Since the underlying consensus engine is agnostic to the application, it only accepts transactions in the form of `[]byte`. The encoding is done by an object called `TxEncoder` and the decoding by an object called `TxDecoder`. 
+Another important use of the Amino is the encoding and decoding of
+[transactions](./transactions.md). Transactions are defined by the application or
+the SDK, but passed to the underlying consensus engine in order to be relayed to
+other peers. Since the underlying consensus engine is agnostic to the application,
+it only accepts transactions in the form of raw bytes. The encoding is done by an
+object called `TxEncoder` and the decoding by an object called `TxDecoder`.
 
 +++ https://github.com/cosmos/cosmos-sdk/blob/7d7821b9af132b0f6131640195326aa02b6751db/types/tx_msg.go#L45-L49
 
 A standard implementation of both these objects can be found in the [`auth` module](https://github.com/cosmos/cosmos-sdk/blob/master/x/auth):
 
 +++ https://github.com/cosmos/cosmos-sdk/blob/7d7821b9af132b0f6131640195326aa02b6751db/x/auth/types/stdtx.go#L241-L266
 
+### Gogoproto
+
+> TODO: Add section on gogoproto semantics and design decisions that are used in the SDK.
+
 ## Next {hide}
```

Comment:
should there be a section on what files need to be fetched for interacting with the modules? Possibly a TODO at least so as modules are transitioned it can be added.

Reply:
I'll add some details here 👍